dataplane/forwarding/fwdport/ports/tap.go

```diff
@@ -129,21 +137,90 @@ func (p *tapPort) Actions(dir fwdpb.PortAction) fwdaction.Actions {
 	return nil
 }
 
-// State return the state of the port (UP).
-// TODO: handle port state correctly.
-func (p *tapPort) State(*fwdpb.PortInfo) (*fwdpb.PortStateReply, error) {
-	ready := fwdpb.PortStateReply{
-		LocalPort: &fwdpb.PortInfo{
-			Laser: fwdpb.PortLaserState_PORT_LASER_STATE_ENABLED,
-		},
-		Link: &fwdpb.LinkStateDesc{
-			State: fwdpb.LinkState_LINK_STATE_UP,
-			RemotePort: &fwdpb.PortInfo{
-				Laser: fwdpb.PortLaserState_PORT_LASER_STATE_ENABLED,
-			},
+func getPortState(name string, ifMgr *kernel.Interfaces) (fwdpb.PortState, fwdpb.PortState, error) {
+	iface, err := ifMgr.LinkByName(name)
+	if err != nil {
+		return fwdpb.PortState_PORT_STATE_UNSPECIFIED, fwdpb.PortState_PORT_STATE_UNSPECIFIED, err
+	}
+	admin, oper := stateFromAttrs(iface.Attrs())
+	return admin, oper, nil
+}
+
+func stateFromAttrs(attrs *netlink.LinkAttrs) (fwdpb.PortState, fwdpb.PortState) {
+	adminState := fwdpb.PortState_PORT_STATE_DISABLED_DOWN
+	if attrs.Flags&net.FlagUp != 0 {
+		adminState = fwdpb.PortState_PORT_STATE_ENABLED_UP
+	}
+
+	var state fwdpb.PortState
+	switch attrs.OperState {
+	case netlink.OperDown:
+		state = fwdpb.PortState_PORT_STATE_DISABLED_DOWN
+	case netlink.OperUp, netlink.OperUnknown: // TAP interface may be unknown state because the dataplane doesn't bind to its fd, so treat unknown as up.
+		state = fwdpb.PortState_PORT_STATE_ENABLED_UP
+	}
+	return adminState, state
+}
+
+func getAndSetState(name string, ifMgr *kernel.Interfaces, pi *fwdpb.PortInfo) (*fwdpb.PortStateReply, error) {
+	adminState, operState, err := getPortState(name, ifMgr)
+	if err != nil {
+		return nil, err
+	}
+	reply := &fwdpb.PortStateReply{
+		Status: &fwdpb.PortInfo{
+			OperStatus:  operState,
+			AdminStatus: adminState,
 		},
 	}
-	return &ready, nil
+	if pi == nil {
+		log.V(1).Infof("dataplane read port %q: admin state %v, oper state %v", name, adminState.String(), operState.String())
+		return reply, nil
+	}
+	log.V(1).Infof("dataplane write port %q: admin state %v, oper state %v, setting to %v", name, adminState.String(), operState.String(), pi.AdminStatus.String())
+	if adminState == pi.GetAdminStatus() {
+		return reply, nil
+	}
+	if pi.AdminStatus == fwdpb.PortState_PORT_STATE_DISABLED_DOWN {
+		err = ifMgr.SetState(name, false)
+	} else if pi.AdminStatus == fwdpb.PortState_PORT_STATE_ENABLED_UP {
+		err = ifMgr.SetState(name, true)
+	}
+	return reply, err
```

Comment:
Here reply.Status is not updated after updating the state. Is this correct? If so can update the doc comment to explain what the behaviour of this function is.